fendermint/vm/topdown/src/finality.rs

```diff
@@ -0,0 +1,479 @@
+        let v = self.parent_view_data.top_down_msgs(height, height)?;
+        Ok(v)
+    }
+
+    fn new_parent_view(
+        &self,
+        height: BlockHeight,
+        block_hash: BlockHash,
+        validator_set: ValidatorSet,
+        top_down_msgs: Vec<CrossMsg>,
+    ) -> StmResult<(), Error> {
+        if !top_down_msgs.is_empty() {
+            // make sure incoming top down messages are ordered by nonce sequentially
+            ensure_sequential_by_nonce(&top_down_msgs)?;
+        };
+
+        let r = self.parent_view_data.height_data.modify(|mut cache| {
+            let r = cache
+                .append(height, (block_hash, validator_set, top_down_msgs))
+                .map_err(Error::NonSequentialParentViewInsert);
+            (cache, r)
+        })?;
+
+        if let Err(e) = r {
+            return abort(e);
+        }
+
+        Ok(())
+    }
+}
+
+impl ParentFinalityProvider for InMemoryFinalityProvider {
+    fn last_committed_finality(&self) -> StmResult<IPCParentFinality, Error> {
+        let finality = self
+            .last_committed_finality
+            .read_clone()?
+            .ok_or(StmError::Abort(Error::CommittedFinalityNotReady))?;
+        Ok(finality)
+    }
+
+    fn next_proposal(&self) -> StmResult<Option<IPCParentFinality>, Error> {
+        let latest_height = if let Some(h) = self.parent_view_data.latest_height()? {
+            h
+        } else {
+            return abort(Error::HeightNotReady);
+        };
+
+        // latest height has not reached, we should wait or abort
+        if latest_height < self.config.chain_head_delay {
+            return abort(Error::HeightThresholdNotReached);
+        }
+
+        let height = latest_height - self.config.chain_head_delay;
+        let last_committed_finality = self.last_committed_finality()?;
+
+        // parent height is not ready to be proposed yet
+        if height <= last_committed_finality.height {
+            return Ok(None);
+        }
+
+        // prepare block hash and validator set
+        let height_data = self.parent_view_data.height_data.read()?;
+        let block_hash = if let Some(v) = height_data.get_value(height) {
+            let (block_hash, _, _) = v;
+            block_hash.clone()
+        } else {
+            return abort(Error::HeightNotFoundInCache(height));
+        };
+
+        Ok(Some(IPCParentFinality { height, block_hash }))
+    }
+
+    fn check_proposal(&self, proposal: &IPCParentFinality) -> StmResult<(), Error> {
+        self.check_height(proposal)?;
+        self.check_block_hash(proposal)
+    }
+
+    fn on_finality_committed(&self, finality: &IPCParentFinality) -> StmResult<(), Error> {
+        // the height to clear
+        let height = finality.height;
+
+        self.parent_view_data.height_data.update(|mut cache| {
+            cache.remove_key_below(height + 1);
+            cache
+        })?;
+
+        self.last_committed_finality.write(Some(finality.clone()))?;
```

Comment:
If this is the only place where this is written, how are you supposed to get the cache started?

The syncer looks like it will get errors as long as  `last_committed_finality` is `None`, and not fetch anything, and therefore it will never propose anything. :chicken: :egg: 

Would be nicer if the `InMemoryFinalityProvider::new` demanded the current finality to be provided.

Reply:
This is a compromise for the circular dependency between provider and App. Maybe making it Optional here is confusing.